File: benchmarks/cycle_churn/cycle_churn.js
// cycle_churn — cyclic-garbage stress test (mirrors cycle_churn.rl).
// Rings are dropped with the cycle intact; V8's tracing GC reclaims them as
// ordinary garbage. total stays below 2^53, so plain numbers are exact.
"use strict";

class Node {
  constructor(value) {
    this.next = null;
    this.value = value;
  }
}

function main() {
  const R = 150000;
  const K = 64;
  let total = 0;

  for (let r = 0; r < R; r++) {
    // Build the ring: first -> ... -> last -> first.
    const first = new Node(r);
    let prev = first;
    for (let i = 1; i < K; i++) {
      const n = new Node(r + i);
      prev.next = n;
      prev = n;
    }
    prev.next = first; // close the cycle

    // Traverse exactly K steps, summing values.
    let cur = first;
    for (let step = 0; step < K; step++) {
      total += cur.value;
      cur = cur.next;
    }
    // Ring dropped here with the cycle intact.
  }

  console.log(total);
}

main();

File: benchmarks/cycle_churn/cycle_churn.py
# cycle_churn — cyclic-garbage stress test (mirrors cycle_churn.rl).
# Rings are dropped with the cycle intact; CPython's refcounting cannot free
# them, so its generational cycle GC does the work — the same division of
# labor as Rolang's ARC + cycle collector.

import sys


class Node:
    __slots__ = ("next", "value")

    def __init__(self, value):
        self.next = None
        self.value = value


def main():
    R = 150000
    K = 64
    total = 0

    for r in range(R):
        # Build the ring: first -> ... -> last -> first.
        first = Node(r)
        prev = first
        for i in range(1, K):
            n = Node(r + i)
            prev.next = n
            prev = n
        prev.next = first  # close the cycle

        # Traverse exactly K steps, summing values.
        cur = first
        for _ in range(K):
            total += cur.value
            cur = cur.next
        # Ring dropped here with the cycle intact.

    sys.stdout.write(f"{total}\n")


if __name__ == "__main__":
    main()
